File: drivers/php/src/Redwire/Frame.php
<?php
/**
 * RedWire frame layout — 16-byte header + payload, all
 * little-endian. Mirrors the engine-side codec
 * (`src/wire/redwire/frame.rs`) and the JS / Java drivers.
 *
 *   u32 length          (whole frame; max 16 MiB)
 *   u8  kind            (one of {@see Frame::KIND_*})
 *   u8  flags           (bit0 = COMPRESSED, bit1 = MORE_FRAMES)
 *   u16 stream_id
 *   u64 correlation_id
 *   payload[length-16]
 *
 * Encoding / decoding go through PHP's binary `pack` / `unpack`. The
 * compression bit is honoured by {@see Codec::compress()} /
 * {@see Codec::decompress()}; the payload field on the value object
 * is always plaintext.
 */

declare(strict_types=1);

namespace Reddb\Redwire;

use Reddb\RedDBException\FrameTooLarge;
use Reddb\RedDBException\ProtocolError;
use Reddb\RedDBException\UnknownFlags;

final class Frame
{
    public const HEADER_SIZE = 16;
    public const MAX_FRAME_SIZE = 16 * 1024 * 1024;
    /** Bits we recognise — anything else trips {@see UnknownFlags}. */
    public const KNOWN_FLAGS = 0b0000_0011;

    /** Magic byte the client writes immediately before the first frame. */
    public const MAGIC = 0xFE;
    /** Highest minor protocol version this driver speaks. */
    public const SUPPORTED_VERSION = 0x01;

    // --- Message kinds (numeric values are part of the wire spec) ---
    public const KIND_QUERY = 0x01;
    public const KIND_RESULT = 0x02;
    public const KIND_ERROR = 0x03;
    public const KIND_BULK_INSERT = 0x04;
    public const KIND_BULK_OK = 0x05;
    public const KIND_BULK_INSERT_BINARY = 0x06;
    public const KIND_QUERY_BINARY = 0x07;
    public const KIND_BULK_INSERT_PREVALIDATED = 0x08;
    public const KIND_HELLO = 0x10;
    public const KIND_HELLO_ACK = 0x11;
    public const KIND_AUTH_REQUEST = 0x12;
    public const KIND_AUTH_RESPONSE = 0x13;
    public const KIND_AUTH_OK = 0x14;
    public const KIND_AUTH_FAIL = 0x15;
    public const KIND_BYE = 0x16;
    public const KIND_PING = 0x17;
    public const KIND_PONG = 0x18;
    public const KIND_GET = 0x19;
    public const KIND_DELETE = 0x1A;
    public const KIND_DELETE_OK = 0x1B;

    // --- Flag bits ---
    public const FLAG_COMPRESSED = 0b0000_0001;
    public const FLAG_MORE_FRAMES = 0b0000_0010;

    public function __construct(
        public readonly int $kind,
        public readonly int $flags,
        public readonly int $streamId,
        public readonly int $correlationId,
        public readonly string $payload,
    ) {
    }

    /** Convenience constructor for the common case (no flags, stream=0). */
    public static function make(int $kind, int $correlationId, string $payload, int $flags = 0): self
    {
        return new self($kind, $flags, 0, $correlationId, $payload);
    }

    public function compressed(): bool
    {
        return ($this->flags & self::FLAG_COMPRESSED) !== 0;
    }

    /**
     * Encode a frame into wire bytes. Honours the COMPRESSED flag.
     * The {@see $payload} field is always plaintext on a Frame value
     * object; this method runs zstd if requested.
     */
    public static function encode(self $frame): string
    {
        $body = $frame->payload;
        $outFlags = $frame->flags & self::KNOWN_FLAGS;
        if (($outFlags & self::FLAG_COMPRESSED) !== 0) {
            try {
                $body = Codec::compress($frame->payload);
            } catch (\Throwable) {
                // Match the engine: drop the flag and ship plaintext.
                $outFlags &= ~self::FLAG_COMPRESSED;
                $body = $frame->payload;
            }
        }
        $bodyLen = strlen($body);
        $total = self::HEADER_SIZE + $bodyLen;
        if ($total > self::MAX_FRAME_SIZE) {
            throw new FrameTooLarge(
                "encoded frame size {$total} exceeds MAX_FRAME_SIZE " . self::MAX_FRAME_SIZE
            );
        }
        // pack: V=u32 LE, C=u8, v=u16 LE, P=u64 LE.
        $header = pack(
            'VCCvP',
            $total,
            $frame->kind & 0xff,
            $outFlags & 0xff,
            $frame->streamId & 0xffff,
            $frame->correlationId,
        );
        return $header . $body;
    }

    /**
     * Decode a complete frame from the start of {@code $bytes}. The
     * buffer must contain at least the declared length; trailing
     * bytes are ignored (use {@see encodedLength()} to slice).
     *
     * @return self plaintext payload (the COMPRESSED flag stays set on the returned object).
     */
    public static function decode(string $bytes): self
    {
        $available = strlen($bytes);
        if ($available < self::HEADER_SIZE) {
            throw new ProtocolError(
                "frame header truncated: got {$available} bytes"
            );
        }
        // unpack returns 1-indexed assoc; ask for named keys.
        $hdr = unpack('Vlength/Ckind/Cflags/vstream/Pcorr', $bytes);
        if ($hdr === false) {
            throw new ProtocolError('frame header: unpack failed');
        }
        $length = $hdr['length'];
        if ($length < self::HEADER_SIZE || $length > self::MAX_FRAME_SIZE) {
            throw new FrameTooLarge("frame length out of range: {$length}");
        }
        if ($available < $length) {
            throw new ProtocolError(
                "frame payload truncated: header says {$length} bytes, only {$available} available"
            );
        }
        $kind = $hdr['kind'];
        $flags = $hdr['flags'];
        if (($flags & ~self::KNOWN_FLAGS) !== 0) {
            throw new UnknownFlags(sprintf('unknown flag bits 0x%02x', $flags));
        }
        $streamId = $hdr['stream'];
        $corr = $hdr['corr'];
        $body = substr($bytes, self::HEADER_SIZE, $length - self::HEADER_SIZE);
        if (($flags & self::FLAG_COMPRESSED) !== 0) {
            $body = Codec::decompress($body);
        }
        return new self($kind, $flags, $streamId, $corr, $body);
    }

    /**
     * Peek the total frame length from a buffer holding at least 4
     * bytes. Used by {@see RedwireConn} to pick the next frame off
     * the read accumulator without copying.
     */
    public static function encodedLength(string $bytes): int
    {
        if (strlen($bytes) < 4) {
            throw new ProtocolError('not enough bytes for length prefix');
        }
        $u = unpack('Vlength', substr($bytes, 0, 4));
        if ($u === false) {
            throw new ProtocolError('failed to unpack length prefix');
        }
        return $u['length'];
    }

    /** Pretty-print a kind byte. Falls back to hex for unknown values. */
    public static function kindName(int $kind): string
    {
        return match ($kind) {
            self::KIND_QUERY => 'Query',
            self::KIND_RESULT => 'Result',
            self::KIND_ERROR => 'Error',
            self::KIND_BULK_INSERT => 'BulkInsert',
            self::KIND_BULK_OK => 'BulkOk',
            self::KIND_BULK_INSERT_BINARY => 'BulkInsertBinary',
            self::KIND_QUERY_BINARY => 'QueryBinary',
            self::KIND_BULK_INSERT_PREVALIDATED => 'BulkInsertPrevalidated',
            self::KIND_HELLO => 'Hello',
            self::KIND_HELLO_ACK => 'HelloAck',
            self::KIND_AUTH_REQUEST => 'AuthRequest',
            self::KIND_AUTH_RESPONSE => 'AuthResponse',
            self::KIND_AUTH_OK => 'AuthOk',
            self::KIND_AUTH_FAIL => 'AuthFail',
            self::KIND_BYE => 'Bye',
            self::KIND_PING => 'Ping',
            self::KIND_PONG => 'Pong',
            self::KIND_GET => 'Get',
            self::KIND_DELETE => 'Delete',
            self::KIND_DELETE_OK => 'DeleteOk',
            default => sprintf('0x%02x', $kind & 0xff),
        };
    }
}

File: drivers/go/redwire/frame.go
// Package redwire encodes / decodes RedWire frames and runs the
// handshake state machine over a TCP / TLS stream. Mirrors
// drivers/rust/src/redwire/ and drivers/js/src/redwire.js — the wire
// shape is fixed by ADR 0001.
package redwire

import (
	"encoding/binary"
	"errors"
	"fmt"
)

// FrameHeaderSize is the fixed 16-byte header that prefixes every frame.
const FrameHeaderSize = 16

// MaxFrameSize is the largest legal encoded frame (header + on-wire payload).
// Anything larger is rejected at decode time so a malformed peer can't allocate
// arbitrary memory on the receiver.
const MaxFrameSize uint32 = 16 * 1024 * 1024

// KnownFlags is the bitmask of every flag the spec defines today. Decoders
// reject any frame that sets bits outside this mask so future flag additions
// fail loudly instead of silently passing.
const KnownFlags uint8 = 0b11

// Flag bits.
const (
	FlagCompressed uint8 = 1 << 0 // 0x01
	FlagMoreFrames uint8 = 1 << 1 // 0x02
)

// MessageKind is the single-byte kind discriminator. Numeric values are part of
// the wire spec — never repurpose a value once shipped.
type MessageKind uint8

// Message kinds. Mirrors src/wire/redwire/frame.rs.
const (
	KindQuery                  MessageKind = 0x01
	KindResult                 MessageKind = 0x02
	KindError                  MessageKind = 0x03
	KindBulkInsert             MessageKind = 0x04
	KindBulkOk                 MessageKind = 0x05
	KindBulkInsertBinary       MessageKind = 0x06
	KindQueryBinary            MessageKind = 0x07
	KindBulkInsertPrevalidated MessageKind = 0x08
	KindHello                  MessageKind = 0x10
	KindHelloAck               MessageKind = 0x11
	KindAuthRequest            MessageKind = 0x12
	KindAuthResponse           MessageKind = 0x13
	KindAuthOk                 MessageKind = 0x14
	KindAuthFail               MessageKind = 0x15
	KindBye                    MessageKind = 0x16
	KindPing                   MessageKind = 0x17
	KindPong                   MessageKind = 0x18
	KindGet                    MessageKind = 0x19
	KindDelete                 MessageKind = 0x1A
	KindDeleteOk               MessageKind = 0x1B
)

// FrameHeader captures the parsed 16-byte header.
type FrameHeader struct {
	Length        uint32 // total frame length incl. header
	Kind          MessageKind
	Flags         uint8
	StreamID      uint16
	CorrelationID uint64
}

// Frame is a decoded RedWire frame. Payload is always the plaintext body — the
// codec inflates compressed frames before delivery so callers never see zstd.
type Frame struct {
	Kind          MessageKind
	Flags         uint8
	StreamID      uint16
	CorrelationID uint64
	Payload       []byte
}

// EncodeFrame serialises the frame to its wire form. Compresses the payload
// when FlagCompressed is set; the COMPRESSED bit on the wire stays so the peer
// knows to decompress.
func EncodeFrame(f *Frame) ([]byte, error) {
	if f == nil {
		return nil, errors.New("redwire: encode nil frame")
	}
	if f.Flags&^KnownFlags != 0 {
		return nil, fmt.Errorf("redwire: encode: unknown flag bits 0x%02x", f.Flags)
	}

	onWire := f.Payload
	flags := f.Flags
	if flags&FlagCompressed != 0 {
		c, err := compressZstd(f.Payload)
		if err != nil {
			// Fallback: ship plaintext, drop the COMPRESSED flag so the peer
			// doesn't try to decompress.
			flags &^= FlagCompressed
		} else {
			onWire = c
		}
	}

	length := uint32(FrameHeaderSize + len(onWire))
	if length > MaxFrameSize {
		return nil, fmt.Errorf("redwire: encode: frame %d > MaxFrameSize %d",
			length, MaxFrameSize)
	}

	buf := make([]byte, length)
	binary.LittleEndian.PutUint32(buf[0:4], length)
	buf[4] = byte(f.Kind)
	buf[5] = flags
	binary.LittleEndian.PutUint16(buf[6:8], f.StreamID)
	binary.LittleEndian.PutUint64(buf[8:16], f.CorrelationID)
	copy(buf[FrameHeaderSize:], onWire)
	return buf, nil
}

// DecodeFrame parses a single frame from the start of buf. Returns the frame
// and the number of bytes consumed (always equal to header.Length on success).
// Compressed frames are inflated before return.
func DecodeFrame(buf []byte) (*Frame, int, error) {
	if len(buf) < FrameHeaderSize {
		return nil, 0, fmt.Errorf("redwire: decode: header truncated (%d < %d)",
			len(buf), FrameHeaderSize)
	}
	length := binary.LittleEndian.Uint32(buf[0:4])
	if length < FrameHeaderSize {
		return nil, 0, fmt.Errorf("redwire: decode: invalid length %d (< %d)",
			length, FrameHeaderSize)
	}
	if length > MaxFrameSize {
		return nil, 0, fmt.Errorf("redwire: decode: invalid length %d (> %d)",
			length, MaxFrameSize)
	}
	if uint32(len(buf)) < length {
		return nil, 0, fmt.Errorf(
			"redwire: decode: payload truncated, expected %d bytes, got %d",
			length, len(buf))
	}
	kindByte := buf[4]
	if !isKnownKind(MessageKind(kindByte)) {
		return nil, 0, fmt.Errorf("redwire: decode: unknown kind 0x%02x", kindByte)
	}
	flags := buf[5]
	if flags&^KnownFlags != 0 {
		return nil, 0, fmt.Errorf("redwire: decode: unknown flag bits 0x%02x", flags)
	}
	streamID := binary.LittleEndian.Uint16(buf[6:8])
	corr := binary.LittleEndian.Uint64(buf[8:16])

	onWire := buf[FrameHeaderSize:length]
	payload := onWire
	if flags&FlagCompressed != 0 {
		plain, err := decompressZstd(onWire)
		if err != nil {
			return nil, 0, fmt.Errorf("redwire: decode: zstd inflate: %w", err)
		}
		payload = plain
	} else {
		// Defensive copy — the input buffer is owned by the reader and gets
		// reused across calls.
		payload = append([]byte(nil), onWire...)
	}
	return &Frame{
		Kind:          MessageKind(kindByte),
		Flags:         flags,
		StreamID:      streamID,
		CorrelationID: corr,
		Payload:       payload,
	}, int(length), nil
}

// NewFrame builds a plaintext frame ready for EncodeFrame.
func NewFrame(kind MessageKind, corr uint64, payload []byte) *Frame {
	return &Frame{
		Kind:          kind,
		CorrelationID: corr,
		Payload:       payload,
	}
}

func isKnownKind(k MessageKind) bool {
	switch k {
	case KindQuery, KindResult, KindError,
		KindBulkInsert, KindBulkOk,
		KindBulkInsertBinary, KindQueryBinary, KindBulkInsertPrevalidated,
		KindHello, KindHelloAck,
		KindAuthRequest, KindAuthResponse, KindAuthOk, KindAuthFail,
		KindBye, KindPing, KindPong,
		KindGet, KindDelete, KindDeleteOk:
		return true
	}
	// Accept the control-plane numeric range that the engine reserves so we
	// don't reject frames the engine might mint (cancel/notice/etc.).
	if k >= 0x20 && k <= 0x27 {
		return true
	}
	return false
}
